src/actions/dropbox/dropbox.ts

```diff
@@ -0,0 +1,187 @@
+import * as querystring from "querystring"
+import * as https from "request-promise-native"
+import {URL} from "url"
+
+import Dropbox = require("dropbox")
+import * as winston from "winston"
+import * as Hub from "../../hub"
+
+export class DropboxAction extends Hub.OAuthAction {
+    name = "dropbox"
+    label = "Dropbox"
+    iconName = "dropbox/dropbox.png"
+    description = "Send query results directly to a file in your Dropbox."
+    supportedActionTypes = [Hub.ActionType.Query, Hub.ActionType.Dashboard]
+    usesStreaming = false
+    minimumSupportedLookerVersion = "6.8.0"
+    requiredFields = []
+    params = []
+
+  async execute(request: Hub.ActionRequest) {
+    const filename = request.formParams.filename
+    const directory = request.formParams.directory
+    const ext = request.attachment!.fileExtension
+
+    let accessToken = ""
+    if (request.params.state_json) {
+      const jsonState = JSON.parse(request.params.state_json)
+      if (jsonState.code && jsonState.redirect) {
+        accessToken = await this.getAccessTokenFromCode(request)
+      }
+    }
+    const drop = this.dropboxClientFromRequest(request, accessToken)
+
+    const resp = new Hub.ActionResponse()
+    resp.success = true
+    if (request.attachment && request.attachment.dataBuffer) {
+      const fileBuf = request.attachment.dataBuffer
+      await drop.filesUpload({path: `/${directory}/${filename}.${ext}`, contents: fileBuf}).catch((err: any) => {
+        winston.error(`Upload unsuccessful: ${JSON.stringify(err)}`)
+        resp.success = false
+        resp.state = new Hub.ActionState()
+        resp.state.data = "reset"
+      })
+    } else {
+      resp.success = false
+      resp.message = "No data sent from Looker to be sent to Dropbox"
+    }
+    return resp
+  }
+
+  async form(request: Hub.ActionRequest) {
+    const form = new Hub.ActionForm()
+    form.fields = []
+
+    let accessToken = ""
+    if (request.params.state_json) {
+      try {
+        const jsonState = JSON.parse(request.params.state_json)
```

Comment:
This parsing appears to happen twice, once here and the other in `getAccessTokenFromCode`. Can that be deduplicated?

Reply:
yeah - I forgot to remove it when I took the key and secret out of the params field